### ISSUE TITLE
system should accept transactions in closed loop, return error during overload

### ISSUE THREAD

**Issue by corecode** [hyperledger-archives/fabric#1098]
Currently the system will run in closed loop mode when using noops consensus, by virtue of how noops is implemented.  The PBFT consensus uses internal data structures and does not limit the number of queued requests.

The system should have a means to:
- delay responses to incoming requests to enable a closed loop feedback
- reject new requests during overload

In any case, once a positive response has been delivered to the client, a message should never be dropped.  This statement requires deliberation on when a positive response can even be delivered; is persistence on stable storage enough?

**Comment by tuand27613** [comment]
@binh @markparz @cca88 @bcbrock

**Comment by cca88** [comment]
Yes: In other words, the consensus layer (aka total-order broadcast channel) can "fill up" when under load. Just like a TCP channel we will need some "flow control" to make it "closed-loop".

Suppose the consensus layer interfaces to a client application with asynchronous SEND and DELIVER events:
+SEND is an input event (to consensus) for submitting a tx/msg for broadcast
+DELIVER is an output event (generated by consensus) that carries an ordered tx/msg

One very simple way to add flow control is via two more events:
+SEND-ACK: an output event from consensus to indicate that it is ready to handle a next SEND event; before that arrives the application must _not_ invoke SEND again
+DELIVER-ACK: an input event (from the application) where the application says that it has consumed the delivered tx/msg and is ready to accept the next one; the consensus will _not_ issue another DELIVER before the application ask invoked DELIVER-ACK

That was implemented in our SINTRA and some of that is described also in the paper --  http://www.zurich.ibm.com/~cca/papers/sintra.pdf

You will not find this PBFT, since it focused on data-center deployment with UDP communication. Under load those packets are dropped and will automatically limit the load.

**Comment by corecode** [comment]
Throughout the stack we will use only bounded queues; request rejection will be handled on the ingress side (REST/Devops).

**Comment by corecode** [comment]
this is a stack-wide issue; the consensus label is incorrect.

**Comment by ibmmark** [comment]
@corecode  are you still working this issue?

**Comment by corecode** [comment]
I have not been working on this - this is a stack-wide issue that needs addressing in many places.

**Comment by mastersingh24** [comment]
@gennadylaventman @bmandler - I would classify this as a flow-control issue we have talked about.  this should fit into the future work once we clean up all of the communication code
